### DISCUSSION TITLE
How can I fix all eslint errors on save?

### DISCUSSION THREAD

**Discussion by mapringg** [LazyVim/LazyVim#402] [Q&A] [answered]
I'm currently using this code and it works, but I feel like there could be a better way:

```
return {
  "neovim/nvim-lspconfig",
  opts = {
    servers = { eslint = {} },
    setup = {
      eslint = function()
        require("lazyvim.util").on_attach(function(client)
          if client.name == "eslint" then
            client.server_capabilities.documentFormattingProvider = true
            vim.api.nvim_create_autocmd("BufWritePre", {
              buffer = bufnr,
              command = "EslintFixAll",
            })
          elseif client.name == "tsserver" then
            client.server_capabilities.documentFormattingProvider = false
          end
        end)
      end,
    },
  },
}
```

I would like to let prettier handle all the formatting and eslint for errors, so this is my null-ls settings:

```
return {
  "jose-elias-alvarez/null-ls.nvim",
  event = { "BufReadPre", "BufNewFile" },
  dependencies = { "mason.nvim" },
  opts = function()
    local nls = require("null-ls")
    return {
      sources = {
        nls.builtins.code_actions.eslint,
        nls.builtins.formatting.prettier,
        nls.builtins.diagnostics.eslint,
        nls.builtins.diagnostics.stylelint,
      },
    }
  end,
}
```

**Comment by folke** [comment, marked as answer]
You should use the eslint lsp server instead of null-ls using this recipe:
https://www.lazyvim.org/configuration/recipes#add-eslint-and-use-it-for-formatting

**Reply by mapringg** [reply]
Thank you for the quick reply. I got it working after I let eslint handles the formatting with [eslint-plugin-prettier](https://github.com/prettier/eslint-plugin-prettier).

**Reply by mapringg** [reply]
With this config, I am not able to format other files such as json with prettier. Is there any workaround for this?

**Reply by folke** [reply]
probably because your eslint config doesn't include files outside your source.

This is what I use in one of my own projects. Might be useful to you:

```js
module. Exports = {
  extends: ["eslint:recommended", "prettier", "plugin:prettier/recommended"],
  plugins: ["prettier"],
  root: true,
  env: {
    node: true,
  },
  rules: {
    "prettier/prettier": "warn",
  },
  overrides: [
    {
      files: ["src/**/*.ts"],
      extends: [
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/strict",
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
        "plugin:unicorn/recommended",
      ],
      plugins: ["@typescript-eslint"],
      parser: "@typescript-eslint/parser",
      parserOptions: {
        tsconfigRootDir: __dirname,
        project: ["./tsconfig.json"],
      },
      rules: {
        "array-callback-return": "error",
        "prefer-template": "warn",
        "prefer-promise-reject-errors": "error",
        "require-unicode-regexp": "error",
        "prefer-spread": "error",
        "lines-between-class-members": [
          "error",
          "always",
          { exceptAfterSingleLine: true },
        ],
        "@typescript-eslint/no-unused-vars": [
          "warn",
          { argsIgnorePattern: "^_" },
        ],
        "unicorn/no-array-for-each": "off",
        "unicorn/prevent-abbreviations": "off",
        "unicorn/explicit-length-check": "off",
        "unicorn/no-array-reduce": "off",
        "unicorn/import-style": "off",
        "unicorn/no-reduce": "off",
        "unicorn/consistent-function-scoping": "off",
      },
    },
  ],
}
```

**Reply by mapringg** [reply]
I meant that file like this cannot be formatted with prettier because I can no longer set prettier as the formatter with null-ls.

<img width="1221" alt="Screenshot 2566-03-10 at 01 03 47" src="https://user-images.githubusercontent.com/15067237/224116031-459602e5-302a-4984-8887-aba5cff37193.png">

**Reply by folke** [reply]
My previous answer was wrong. I also use jsonls and use its formatter for json.
You could still use prettier with null-ls for json if you want. Just add it as you normally would and configure it to only run for json files. Check the null-ls docs, but something like this:

```lua
          nls.builtins.formatting.prettier.with({
            filetypes = { "json" }, -- only runs `prettier` for json
          }),
```

**Reply by mapringg** [reply]
Thank you. This is exactly what I want. I started using neovim a few days ago and lazyvim made the transition very smooth.

**Comment by ooloth** [comment]
The original question was about automatically fixing `eslint` errors, rather than using `eslint` as a formatter.

@folke Your answer works if you don't use `prettier` (or you integrate `prettier` with `eslint`), but otherwise it unnecessarily breaks `prettier`'s functionality.

If you want to scope `eslint`'s behavior to automatically fixing linting errors only (and allow `prettier` to continue handling formatting), you can also use this approach:

```lua
-- lua/plugins/your-lsp-settings-file.lua

return {
{
  'neovim/nvim-lspconfig',
  -- other settings removed for brevity
  opts = {
    servers = {
      eslint = {
        settings = {
          -- helps eslint find the eslintrc when it's placed in a subfolder instead of the cwd root
          workingDirectory = { mode = 'auto' },
        },
      },
    setup = {
      eslint = function()
        -- automatically fix linting errors on save (but otherwise do not format the document)
        vim.cmd([[
          autocmd BufWritePre *.tsx,*.ts,*.jsx,*.js EslintFixAll
        ]])
      end,
    },
  }
}

-- no need for the `null-ls` version of `eslint`
```

**Reply by mapringg** [reply]
This will definitely be helpful when I do not want eslint to handle formatting.

**Comment by folke** [comment]
@ooloth's suggestion is actually the [recommended way](https://prettier.io/docs/en/integrating-with-linters.html) to integrate prettier with linters.

I've added two extras for eslint and one for prettier and updated the docs
https://www.lazyvim.org/configuration/recipes#use-eslint-for-fix-on-save-and-prettier-for-formatting

**Reply by mapringg** [reply]
Thank you for updating the docs.
